pkg/gofr/datasource/file/ftp/fs.go

```diff
@@ -0,0 +1,276 @@
+
+	err := f.conn.Stor(filePath, emptyReader)
+	if err != nil {
+		return nil, err
+	}
+
+	f.logger.Logf("Created %s", name)
+
+	s := strings.Split(filePath, "/")
+
+	res, err := f.conn.Retr(filePath)
+	if err != nil {
+		return nil, err
+	}
+
+	defer res.Close()
+
+	return &ftpFile{
+		response: res,
+		name:     s[len(s)-1],
+		path:     filePath,
+		conn:     f.conn,
+		logger:   f.logger,
+	}, nil
+}
+
+// Mkdir creates a directory on the FTP server. Here, os.FileMode is unused.
+func (f *ftpFileSystem) Mkdir(name string, _ os.FileMode) error {
+	if name == "" {
+		return errEmptyDirectory
+	}
+
+	filePath := fmt.Sprintf("%s/%s", f.config.RemoteDir, name)
+
+	err := f.conn.MakeDir(filePath)
+	if err == nil {
+		f.logger.Logf("Created %s directory", name)
+		return nil
+	}
+
+	return err
+}
+
+// MkdirAll creates directories recursively on the FTP server. Here, os.FileMode is unused.
+func (f *ftpFileSystem) MkdirAll(path string, _ os.FileMode) error {
+	dirs := strings.Split(path, "/")
+
+	currentDir := dirs[0]
+
+	v := 0
+
+	for i, dir := range dirs {
+		// Ignore empty directory names (can happen if there are double slashes).
+		if dir == "" {
+			continue
+		}
+
+		if i == 0 {
+			currentDir = dir
+		} else {
+			currentDir = fmt.Sprintf("%s/%s", currentDir, dir)
+		}
+
+		err := f.conn.MakeDir(currentDir)
+		if err != nil {
+			// if error indicates that directory exists continue, else return.
+			if fmt.Sprint(err) != directoryAlreadyExistError {
+				continue
+			}
+
+			return err
+		}
+
+		v++
+	}
+
+	if v == 0 {
+		return errors.New("Error Creating Directory")
+	}
+
+	f.logger.Logf("Created directories with path %s", path)
+
+	return nil
+}
+
+// Open retrieves a file from the FTP server and returns a file handle.
+func (f *ftpFileSystem) Open(name string) (File, error) {
+	if name == "" {
+		return nil, errEmptyFilename
+	}
+
+	filePath := fmt.Sprintf("%s/%s", f.config.RemoteDir, name)
+
+	res, err := f.conn.Retr(filePath)
+	if err != nil {
+		return nil, err
+	}
+
+	s := strings.Split(filePath, "/")
+
+	f.logger.Logf("Opened %s", name)
+
+	defer res.Close()
+
+	return &ftpFile{
+		response: res,
+		name:     s[len(s)-1],
+		path:     filePath,
+		conn:     f.conn,
+		logger:   f.logger,
+	}, nil
+}
+
+// permissions are not clear for Ftp as file commands do not accept an argument and don't store their file permissions.
+// currently, this function just calls the Open function. Here, os.FileMode is unused.
+func (f *ftpFileSystem) OpenFile(name string, _ int, _ os.FileMode) (File, error) {
+	return f.Open(name)
+}
+
+// Remove deletes a file from the FTP server.
+func (f *ftpFileSystem) Remove(name string) error {
+	if name == "" {
+		return errEmptyFilename
+	}
+
+	filePath := fmt.Sprintf("%s/%s", f.config.RemoteDir, name)
+
+	err := f.conn.Delete(filePath)
+	if err == nil || fmt.Sprint(err) == transferComplete {
+		f.logger.Logf("%s successfully removed", name)
+
+		err = nil
```

Comment:
There’s not really any good reason to assign `nil` into `err` so someone else can return it, when we can just `return nil` here ourselves.

Reply:
Noted